runtime/common/src/paras_registrar.rs

```diff
@@ -0,0 +1,681 @@
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Module to handle parathread/parachain registration and related fund management.
+//! In essence this is a simple wrapper around `paras`.
+
+use sp_std::{prelude::*, result};
+
+use frame_support::{
+	decl_storage, decl_module, decl_error, ensure,
+	dispatch::DispatchResult,
+	traits::{Get, Currency, ReservableCurrency},
+	weights::DispatchClass,
+};
+use frame_system::{self, ensure_root, ensure_signed};
+use primitives::v1::{
+	Id as ParaId, ValidationCode, HeadData,
+};
+use runtime_parachains::{
+	paras::{
+		self,
+		ParaGenesisArgs,
+	},
+	ensure_parachain,
+	Origin,
+};
+
+type BalanceOf<T> =
+	<<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::Balance;
+
+pub trait Trait: paras::Trait {
+	/// The aggregated origin type must support the `parachains` origin. We require that we can
+	/// infallibly convert between this origin and the system origin, but in reality, they're the
+	/// same type, we just can't express that to the Rust type system without writing a `where`
+	/// clause everywhere.
+	type Origin: From<<Self as frame_system::Trait>::Origin>
+		+ Into<result::Result<Origin, <Self as Trait>::Origin>>;
+
+	/// The system's currency for parathread payment.
+	type Currency: ReservableCurrency<Self::AccountId>;
+
+	/// The deposit to be paid to run a parathread.
+	type ParathreadDeposit: Get<BalanceOf<Self>>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Registrar {
+		/// Pending swap operations.
+		PendingSwap: map hasher(twox_64_concat) ParaId => Option<ParaId>;
+
+		/// Map of all registered parathreads/chains.
+		Paras get(fn paras): map hasher(twox_64_concat) ParaId => Option<bool>;
+
+		/// Users who have paid a parathread's deposit.
+		Debtors: map hasher(twox_64_concat) ParaId => T::AccountId;
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// Parachain already exists.
+		ParaAlreadyExists,
+		/// Invalid parachain ID.
+		InvalidChainId,
+		/// Invalid parathread ID.
+		InvalidThreadId,
+		/// Invalid para code size.
+		CodeTooLarge,
+		/// Invalid para head data size.
+		HeadDataTooLarge,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: <T as frame_system::Trait>::Origin {
+		type Error = Error<T>;
+
+		/// Register a parathread with given code for immediate use.
+		///
+		/// Must be sent from a Signed origin that is able to have `ParathreadDeposit` reserved.
+		/// `gensis_head` and `validation_code` are used to initalize the parathread's state.
+		#[weight = (5_000_000_000, DispatchClass::Operational)]
+		pub fn register_parathread(
```

Comment:
These calls may not even need to be calls, per-se. They could just be free functions that are invoked by e.g. the `slots` module